addons/sale_stock_margin/models/sale_order_line.py

```diff
@@ -14,17 +14,16 @@ def _compute_purchase_price(self):
             product = line.product_id.with_company(line.company_id)
             if not line.has_valued_move_ids():
                 line_ids_to_pass.add(line.id)
-            elif (
+            elif line.product_id and line.product_id.categ_id.property_cost_method != 'standard':
                 # don't overwrite any existing value unless non-standard cost method
-                (line.product_id and line.product_id.categ_id.property_cost_method != 'standard') or
-                # if line added from delivery, allow recomputation
-                (not line.product_uom_qty and line.qty_delivered)
-            ):
                 purch_price = product._compute_average_price(0, line.product_uom_qty or line.qty_to_invoice, line.move_ids)
```

Comment:
In your test, the product you add on the picking has `invoice_policy == 'order'`. Meaning `line.qty_to_invoice == line.product_uom_qty == 0` in the call to `_compute_average_price()` and thus returns `0`.
To me it's still mandatory to get the price from `_compute_average_price` in case the product cost changed after the delivery validation. 
To get the correct value, we should pass `qty_delivered` to `_compute_average_price` in case of `invoice_policy == 'delivery'` products